accepted/3.0/patterns/exhaustiveness.md

```diff
@@ -417,39 +417,49 @@ The lifted space union for a pattern with matched value type `M` is:
     exhaustiveness algorithm doesn't model "all objects that are not `int`", so
     it can't tell that this is exhaustive.*
 
-    *But with sealed types, exhaustiveness can sometimes represent the set of
-    remaining types. Given:*
+    *However, we can have a situation where a cast pattern is used to ensure
+    that a case "handles" all possible matched values because it will match
+    or it will throw.*
 
     ```dart
-    sealed class A {
-      final int field;
-    }
-    class B extends A {}
-    class C extends A {}
-    class D extends A {}
-
-    test(A a) => switch (a) {
-      C(field: 0) as C => 0,
-      C _ => 1
+    abstract class A {}
+    class B implements A {}
+    class C implements A {}
+    class D implements A {}
+
+    test(A a) => switch (a) { // `a` should be a `B` or a `C`.
+      B() => 0,
+      _ as C => 1,
     };
     ```
 
-    *After the first case, we know that if `a` is a `B` or `D` then we will have
-    thrown an exception. But if `a` is a `C`, it may not have matched if the
-    field isn't `0`. So the space for the first case is a union of `B|C(field:
-    0)|D'.*
-
-    *Then the second case covers the rest of `C` and this is exhaustive.*
+    *In this example, `test` should deal with instances of `B` and instances of
+    `C`, but it should simply throw if `a` has any other type. So we use the
+    cast pattern in the last case to ensure that the switch is exhaustive, which
+    is true because all instances of `B` will be handled by the subpattern, and
+    all other objects will cause the cast to throw.*
 
     Formally, the space union `spaces` for a cast pattern with cast type `C` is
-    a union of:
+    computed as follows:
+
+    Let `S` be the lifted space union of the cast's subpattern in context `C`.
+
+    1.  If `C` is a *subset* (see below) of `S` then the result is the lifted
```

Comment:
**subset** is not defined for a type and a space and the **space subsetting** definition is not adequate for our purpose here. 

A possible definition could be:

---
***A) Type subset of space union***
A type `C` is a **subset** of space union `S` iff

Either
1. `C` is **subset** of each single space `S_i` of `S` (see below), or
2. for each **subtype** `C_i` of `C` there exists a single space `S_i` of `S` such that `C_i` is a subset of `S_i`.

***B) Type subset of single space***
A type `C` is a **subset** of single space `S` iff
1. `C` is a subtype of the type `T` of `S`, 
2. `S` has an open restriction, and
3. For each property `p_j` of `S_i` with space `S_j` and static type `C_j`, `C_j` is a subset of space `S_j`.

***C) Subtypes of type***
The **subtypes** of a type `C` is
1. If `C` is sealed, the expanded subtypes is the enumerated subtypes of `C`.
2. Otherwise the singleton set `{C}`.
---
I'm sure we have a definition of "enumerated subtypes", though, but it is meant to be the subtypes found in the section "The declaration D of T is a sealed type" of [Expanding a space](https://github.com/dart-lang/language/blob/main/accepted/3.0/patterns/exhaustiveness.md#expanding-a-space) algorithm.

The motivation for these rules are:
```
sealed class M {}
class A extends M {}
sealed class B<T> extends M {}
class C<T> extends B<T> {}
class D<T, S> extends B<T> {}

method1(o) => switch (o) {
    (B() || B()) as B => 0, // Rules A.1, B.1 (is subset)
  };
method2(o) => switch (o) {
    (A() || B()) as M => 0, // Rules A.2, B.1, C.1 (is subset)
  };
method3(o) => switch (o) {
    [_] as List, // Rule B.2, (not subset)
    [...] as List, // Rule B.2, (is subset)
  };
method4(o) => switch (o) {
    int(isEven: true) as int, // Rules A.1, B.3, C.2 (not subset)
    int(: var isEven) as int, // Rules A.1, B.3, C.2 (is subset)
}
```
Do to the limitations of the "subtypes" algorithm used in C.1 we don't find this to be exhausting:
```
method5(o) => s
    (A() || C() || D()) as M => 0, // Rules A.2, B.1, C.1 (transitively)
  };
```
but this similar to this, which also don't find exhausting:
```
method6(M o) => s
    A() => 0,
    C() => 0,
    D() => 0,
  };
```

Reply:
@johnniwinther and I have discussed this topic today. I'll wait a bit, and then try to come up with a revised PR when the matter has stabilized.

Reply:
Thanks, I used this in the new commit (but with several changes). PTAL!